agent/acs/session/task_stop_verification_ack_responder.go

```diff
@@ -39,6 +39,7 @@ func (ts *taskStopper) StopTask(taskARN string) {
 			loggerfield.TaskARN: task.Arn,
 		})
 		task.SetDesiredStatus(apitaskstatus.TaskStopped)
+		task.UpdateDesiredStatus()
 		ts.taskEngine.AddTask(task)
```

Comment:
I am wondering what is the point here of the `AddTask()` call. If you follow the code path, for a task that exists in the engine, this method is essentially a nop, especially since the task and all of its containers' desired states have already been updated to STOPPED. 

wdyt about getting rid of that (line 43)?

Reply:
Agreed that L41-L42 and L43 are redundant alongside each other. 

What do you think about getting rid of L41 and L42 instead? That way, we let the task manager be responsible for stopping tasks (which is what the ACS payload message responder does) and not have this fork in logic in how tasks are stopped. L43 also ensures that task state is saved in Agent's local state DB when desired status changes.

Reply:
I agree. All the other places we call `AddTask()` are when we ingest a new task. Therefore it doesn't seem to be the right call here and might incur unintended side effects (if `AddTask() has any assumptions of the task being "new"). 

I do want to make sure that Agent is able to gracefully handle the case where the task somehow is not known. Although this should not be possible technically because Agent has to put a task into `TaskStopVerification` in the first place, for the task to show up in `TaskStopVerificationAck`, might still be a good idea to have a check/filter.

Reply:
> What do you think about getting rid of L41 and L42 instead?

`AddTask` only updates the task state if it's a task that agent hasn't seen yet (or has been removed from its state engine): https://github.com/aws/amazon-ecs-agent/blob/0a1a88d83d7ee34c593f663979b8abd920f55481/agent/engine/docker_task_engine.go#L1184

So, with just line 43, I think we still have a leak. 

If we can update the desired status for the task and just drop it on the ACS channel to handle it, then I think your line of reasoning holds dane. Otherwise, I prefer 41-42 over 43 (and getting rid of it).

Reply:
> All the other places we call AddTask() are when we ingest a new task. 

Could you clarify what you mean by "new task" here? My understanding is that `AddTask` does get called for tasks that Agent knows about (e.g., Agent receives an ACS payload message to stop a task that is running).

> I do want to make sure that Agent is able to gracefully handle the case where the task somehow is not known.

Following the code path for`AddTask`, this is actually already done by the task engine, see: https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/docker_task_engine.go#L2646-L2652

Reply:
@aaithal 

> If we can update the desired status for the task and just drop it on the ACS channel to handle it, then I think your line of reasoning holds dane

Perhaps I may be misunderstanding the code, but isn't this already done? If `exists` evaluates to true in the line of code that you linked, then this line of code still gets executed:
https://github.com/aws/amazon-ecs-agent/blob/0a1a88d83d7ee34c593f663979b8abd920f55481/agent/engine/docker_task_engine.go#L1206

Following that code path emits the desired status for the task on the ACS channel here: https://github.com/aws/amazon-ecs-agent/blob/0a1a88d83d7ee34c593f663979b8abd920f55481/agent/engine/docker_task_engine.go#L2658-L2664

Reply:
Originally I was thinking that `AddTask()` should only be used for new task ingestion, but like you said, payload message handler calls it for all tasks, stopping or starting ([code](https://github.com/aws/amazon-ecs-agent/blob/master/agent/acs/session/payload_responder.go#L253)), which has been working fine, so that means AddTask() is able to handle a task that has an existing status of RUNNING/RUNNING, but a new status of RUNNING/STOPPED.

I'm fine with removing L41-42 in favor of L43

Reply:
Yeah, I think you're right. I missed line 1206 there. If we think `AddTask()` means that we have one single way through which agent is applying the stop logic, then I'm fine with it. Can you test it once though with the repro scenario (remove 41-42 and keep 43 around)?

Reply:
@danehlim How does [UpdateDesiredStatus](https://github.com/aws/amazon-ecs-agent/blob/0a1a88d83d7ee34c593f663979b8abd920f55481/agent/api/task/task.go#L2620), which updates the container to stopped, get called in that code path? If we go with L43 route?

Reply:
> Can you test it once though with the repro scenario (remove 41-42 and keep 43 around)?

Will do, the plan is to make sure any changes to the initial revision of this pull request are tested against internal bug repro scenario before merging this pull request. I will also update the pull request description to include that this testing is done.

Reply:
Thanks much for the discussion @aaithal and @yinyic !

Reply:
L41 is actually necessary otherwise Agent may not know that task needs to be stopped. Discussed and aligned offline with @aaithal and @yinyic that we will go with removing L43 and keeping L41-L42. In addition to this, we will add logic to save task state to local state DB when its status is updated to STOPPED, since the AddTask call (which includes saving to local state DB logic) is being removed.